rescan.go

```diff
@@ -476,6 +484,51 @@ func rescan(chain ChainSource, options ...RescanOption) error {
 				return err
 			}
 		}
+
+		return nil
+	}
+
+	// We'll need to ensure that the backing chain has actually caught up to
+	// the rescan's starting height. If it hasn't, we'll subscribe for block
+	// notifications at tip and wait until we receive a notification for a
+	// block with the rescan's starting height.
+	log.Debugf("Waiting to catch up to the rescan start height=%d",
+		curStamp.Height)
+
+	if err := waitFor(func(_ chainhash.Hash, height uint32) bool {
+		return height >= uint32(curStamp.Height)
+	}); err != nil {
+		return err
+	}
+
+	// To ensure that we batch as many filter queries as possible, we also
+	// wait for the header chain to either be current or for the .
+	r, ok := chain.(*RescanChainSource)
+	if ok {
+		log.Debugf("Waiting for the chain source to be current or " +
+			"for the rescan end height to be reached.")
+
+		if err := waitFor(func(hash chainhash.Hash, height uint32) bool {
+			// If the header chain is current, then there
+			// is no need to wait.
+			if r.IsCurrent() {
+				return true
+			}
+
+			// If an end height was specified then we wait until
+			// the notification corresponding to that block height.
+			if ro.endBlock.Height > 0 &&
+				height >= uint32(ro.endBlock.Height) {
+
+				return true
+			}
+
+			// If a block hash was specified, check if the
+			// notification is for that block.
+			return hash == ro.endBlock.Hash
+		}); err != nil {
+			return err
+		}
```

Comment:
could maybe change this a bit so that if we are lagging by at least 1000, we start preemptively fetching filters (similarly to what is done for filter header syncing)